boards/qemu_i486_q35/src/io.rs

```diff
@@ -2,15 +2,51 @@
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 // Copyright Tock Contributors 2024.
 
-use core::panic::PanicInfo;
+use core::fmt::Write;
 use core::ptr;
+use core::{arch::asm, panic::PanicInfo};
 
 use kernel::debug;
 
 use x86_q35::serial::{BlockingSerialPort, COM1_BASE};
 
 use crate::{CHIP, PROCESSES, PROCESS_PRINTER};
 
+/// Exists QEMU
+///
+/// This function requires the `-device isa-debug-exit,iobase=0xf4,iosize=0x04`
+/// device enabled.
+fn exit_qemu() -> ! {
+    unsafe {
+        asm!(
+            "
+        mov dx, 0xf4
+        mov al, 0x01
+        out dx,al
+        "
+        );
+    }
+
+    // We prefer the infinite loop to the `options(noreturn)` for `asm!` as
+    // the required `isa-debug-exit` device might be missing in which case
+    // the execution does not stop and generates undefined behaviour.
+    let mut com1 = unsafe { BlockingSerialPort::new(COM1_BASE) };
+    let _ = com1.write_fmt(format_args!(
+        "BUG:  QEMU did not exit.\
+        \r\n      The isa-debug-exit device is missing or is at a wrong address.\
+        \r\n      Please make sure the QEMU command line uses\
+        \r\n      the `-device isa-debug-exit,iobase=0xf4,iosize=0x04` argument.\
+        \r\nHINT: Use `killall qemu-system-i386` or the Task Manager to stop.\
```

Comment:
```suggestion
        \r\nHINT: Use the QEMU monitor, `killall qemu-system-i386` or the Task Manager to stop QEMU.\
```

Reply:
I don't know what the qemu monitor is :)